### DISCUSSION TITLE
null, optional, 널 객체 특징을 설명하고 어떤 방식을 선호하는지 근거와 함께 알려주세요.

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#100] [Clean Code]
.

**Comment by J-I-H-O** [comment]
**null**
null은 NPE가 발생할 수 있는 위험이 있습니다. 따라서 코드 안정성을 저하시킬 수 있고, null 처리를 위한 추가적인 로직이 필요합니다.

**Optional**
자바 8부터 도입된 클래스로, 값이 없을 수 있음을 나타낼 수 있는 방법 중 하나입니다.
명시적으로 값이 없음을 나타낼 수 있으므로 NPE를 방지할 수 있습니다.
메서드 체인을 통해 값을 안전하게 처리할 수 있는 기능을 제공합니다.

**널 객체(Null Object)**
null 대신 특별한 객체를 반환하는 방법입니다. 예외 처리나 null 체크를 줄일 수 있습니다.

저는 널 객체를 사용하는 방법을 선호합니다. 
null을 직접 사용하는 경우 NPE가 발생할 수 있는 위험성이 존재하고, null 처리를 위한 추가적인 코드가 필요합니다.
Optional은 null 처리를 위한 메서드를 제공하지만, 가독성이 떨어지고 잘못 사용하는 경우 코드가 불필요하게 복잡해진다는 단점이 있습니다.
널 객체를 사용하는 경우, 명시적으로 예외 상황임을 나타낼 수 있는 동시에 NPE도 방지할 수 있고, 가독성도 좋기 때문에 다른 방법에 비해 선호합니다.

**Comment by 3Juhwan** [comment]
null은 자바에서 객체가 아무런 값을 가지지 않음을 나타내는 값을 의미합니다. null을 사용하면 NullPointerException 등의 예외를 발생시킬 수 있는 위험성이 있습니다.

Optional은 자바 8부터 도입된 클래스로, 값이 존재할 수도 있고 존재하지 않을 수도 있는 상황을 나타내기 위해 사용됩니다. Optional은 값이 존재하는 경우에만 값을 반환하고, 값이 없는 경우에는 예외를 발생시키지 않고 안전하게 처리할 수 있도록 도와줍니다.

널 객체 패턴은 null 대신에 객체가 아무런 값을 가지지 않음을 나타낼 수 있습니다. 이를 통해 NullPointerException을 방지하고 안전하게 코드를 작성할 수 있습니다. 또한, 널 객체 패턴을 사용하면 코드의 가독성을 높일 수 있습니다. 널 객체 패턴은 값이 없는 경우에도 명시적인 객체를 반환하므로, 코드를 이해하고 디버깅하기 쉽습니다. 저는 이 이유로 널 객체 패턴을 선호합니다.
